File: packages/easy_compressor/ios/Classes/EasyCompressorPlugin.swift
import Flutter
import UIKit

public class EasyCompressorPlugin: NSObject, FlutterPlugin {
    private var compressor: VideoCompressor?
    private var progressSink: FlutterEventSink?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = EasyCompressorPlugin()

        let channel = FlutterMethodChannel(
            name: "easy_compressor",
            binaryMessenger: registrar.messenger()
        )
        registrar.addMethodCallDelegate(instance, channel: channel)

        let progressChannel = FlutterEventChannel(
            name: "easy_compressor/progress",
            binaryMessenger: registrar.messenger()
        )
        progressChannel.setStreamHandler(instance)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "compressVideo":
            guard let args = call.arguments as? [String: Any],
                  let inputPath = args["inputPath"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing inputPath", details: nil))
                return
            }

            let quality = args["quality"] as? Int ?? 70
            let maxHeight = args["maxHeight"] as? Int
            let maxWidth = args["maxWidth"] as? Int
            let frameRate = args["frameRate"] as? Int
            let includeAudio = args["includeAudio"] as? Bool ?? true
            let audioBitrate = args["audioBitrate"] as? Int ?? 128000
            let videoCodec = args["videoCodec"] as? String ?? "h264"
            let outputPath = args["outputPath"] as? String

            compressor = VideoCompressor()
            compressor?.compress(
                inputPath: inputPath,
                quality: quality,
                maxHeight: maxHeight,
                maxWidth: maxWidth,
                frameRate: frameRate,
                includeAudio: includeAudio,
                audioBitrate: audioBitrate,
                videoCodec: videoCodec,
                outputPath: outputPath,
                onProgress: { [weak self] progress in
                    DispatchQueue.main.async {
                        self?.progressSink?(progress)
                    }
                },
                completion: { compResult in
                    DispatchQueue.main.async {
                        switch compResult {
                        case .success(let data):
                            result(data)
                        case .failure(let error):
                            result(FlutterError(
                                code: "COMPRESSION_FAILED",
                                message: error.localizedDescription,
                                details: nil
                            ))
                        }
                    }
                }
            )

        case "getMediaInfo":
            guard let args = call.arguments as? [String: Any],
                  let inputPath = args["inputPath"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing inputPath", details: nil))
                return
            }

            DispatchQueue.global(qos: .userInitiated).async {
                do {
                    let info = try MediaInfoExtractor.extract(path: inputPath)
                    DispatchQueue.main.async { result(info) }
                } catch {
                    DispatchQueue.main.async {
                        result(FlutterError(
                            code: "MEDIA_INFO_FAILED",
                            message: error.localizedDescription,
                            details: nil
                        ))
                    }
                }
            }

        case "getThumbnail":
            guard let args = call.arguments as? [String: Any],
                  let inputPath = args["inputPath"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing inputPath", details: nil))
                return
            }

            let positionMs = args["positionMs"] as? Int ?? 0
            let quality = args["quality"] as? Int ?? 80
            let maxHeight = args["maxHeight"] as? Int

            DispatchQueue.global(qos: .userInitiated).async {
                let data = MediaInfoExtractor.getThumbnail(
                    path: inputPath,
                    positionMs: positionMs,
                    quality: quality,
                    maxHeight: maxHeight
                )
                DispatchQueue.main.async {
                    result(data != nil ? FlutterStandardTypedData(bytes: data!) : nil)
                }
            }

        case "cancelCompression":
            compressor?.cancel()
            result(nil)

        case "clearCache":
            VideoCompressor.clearCache()
            result(nil)

        default:
            result(FlutterMethodNotImplemented)
        }
    }
}

extension EasyCompressorPlugin: FlutterStreamHandler {
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        progressSink = events
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        progressSink = nil
        return nil
    }
}

File: packages/easy_compressor/android/src/main/kotlin/com/example/easy_compressor/EasyCompressorPlugin.kt
package com.example.easy_compressor

import android.content.Context
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.*

class EasyCompressorPlugin : FlutterPlugin, MethodCallHandler {
    private lateinit var channel: MethodChannel
    private lateinit var progressChannel: EventChannel
    private lateinit var context: Context
    private var progressSink: EventChannel.EventSink? = null
    private var compressor: VideoCompressor? = null
    private val scope = CoroutineScope(Dispatchers.Main + SupervisorJob())

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        channel = MethodChannel(binding.binaryMessenger, "easy_compressor")
        channel.setMethodCallHandler(this)

        progressChannel = EventChannel(binding.binaryMessenger, "easy_compressor/progress")
        progressChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                progressSink = events
            }

            override fun onCancel(arguments: Any?) {
                progressSink = null
            }
        })
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "compressVideo" -> {
                val inputPath = call.argument<String>("inputPath")!!
                val quality = call.argument<Int>("quality") ?: 70
                val maxHeight = call.argument<Int>("maxHeight")
                val maxWidth = call.argument<Int>("maxWidth")
                val frameRate = call.argument<Int>("frameRate")
                val includeAudio = call.argument<Boolean>("includeAudio") ?: true
                val audioBitrate = call.argument<Int>("audioBitrate") ?: 128000
                val videoCodec = call.argument<String>("videoCodec") ?: "h264"
                val outputPath = call.argument<String>("outputPath")

                scope.launch {
                    try {
                        compressor = VideoCompressor(context)
                        val compResult = compressor!!.compress(
                            inputPath = inputPath,
                            quality = quality,
                            maxHeight = maxHeight,
                            maxWidth = maxWidth,
                            frameRate = frameRate,
                            includeAudio = includeAudio,
                            audioBitrate = audioBitrate,
                            videoCodec = videoCodec,
                            outputPath = outputPath,
                            onProgress = { progress ->
                                scope.launch(Dispatchers.Main) {
                                    progressSink?.success(progress)
                                }
                            }
                        )
                        result.success(compResult)
                    } catch (e: Exception) {
                        result.error("COMPRESSION_FAILED", e.message, e.toString())
                    }
                }
            }

            "getMediaInfo" -> {
                val inputPath = call.argument<String>("inputPath")!!
                scope.launch {
                    try {
                        val info = MediaInfoExtractor.extract(context, inputPath)
                        result.success(info)
                    } catch (e: Exception) {
                        result.error("MEDIA_INFO_FAILED", e.message, e.toString())
                    }
                }
            }

            "getThumbnail" -> {
                val inputPath = call.argument<String>("inputPath")!!
                val positionMs = call.argument<Int>("positionMs") ?: 0
                val quality = call.argument<Int>("quality") ?: 80
                val maxHeight = call.argument<Int>("maxHeight")

                scope.launch {
                    try {
                        val bytes = MediaInfoExtractor.getThumbnail(
                            context, inputPath, positionMs.toLong(), quality, maxHeight
                        )
                        result.success(bytes)
                    } catch (e: Exception) {
                        result.error("THUMBNAIL_FAILED", e.message, e.toString())
                    }
                }
            }

            "cancelCompression" -> {
                compressor?.cancel()
                result.success(null)
            }

            "clearCache" -> {
                scope.launch {
                    try {
                        VideoCompressor.clearCache(context)
                        result.success(null)
                    } catch (e: Exception) {
                        result.error("CLEAR_CACHE_FAILED", e.message, e.toString())
                    }
                }
            }

            else -> result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        scope.cancel()
    }
}
